### PR TITLE
feat(skin): inverse-distance auto skin weights (closes #402)

### LINKED ISSUE #402: AI: libigl bounded biharmonic skinning weights
Parent epic: #397

## Goal

Compute high-quality skinning weights for a mesh + skeleton using **libigl**'s bounded biharmonic weights — no ML, header-only C++.

## Scope

- Add libigl as a header-only dep (only the subset we need: `igl::bbw`, `igl::boundary_conditions`).
- CLI: `qtmesh skin model.fbx --algo biharmonic -o skinned.fbx`.
- GUI: Animation Mode action "Compute skin weights" when an entity has a skeleton but missing/poor weights.
- MCP: `compute_skin_weights` tool.
- Pair with the auto-rigger task (#397 Tier 3) — biharmonic weights are the natural completion step after Pinocchio places the skeleton.
- Preview: re-render the entity with new weights and a default test pose.

## Acceptance Criteria

- [ ] A static mesh + a manually placed skeleton produces deformable skinning that follows bone rotations smoothly.
- [ ] Weights sum to 1.0 per vertex and are non-negative.
- [ ] Works for at least the standard humanoid skeleton from the test suite.
- [ ] Unit test on a simple mesh+skeleton fixture.
- [ ] Sentry breadcrumb `ai.assist.skin_weights` recorded.

## Effort

~5 days.

### PR DESCRIPTION
**Pull request by fernandotonon**
## Summary

Closes #402 (epic #397, AI-Assist slice 5). Adds automatic skin weight computation for skinned meshes.

## Background — why not libigl BBW

The issue proposes wrapping libigl's bounded biharmonic weights (BBW). In practice:

- BBW solves a biharmonic equation over the **volume** of the mesh — requires tetrahedralization via TetGen.
- TetGen is **GPL/copyleft**. Linking it would force the entire binary to GPL and close off Homebrew / Snap / WinGet redistribution under the project's permissive-license stance.
- TetGen meshing also fails on common asset issues (non-manifold, self-intersections, degenerate tris) — Mixamo FBX would routinely fail before BBW even runs.

This slice ships an **inverse-distance heuristic with zero new dependencies**. The `Algorithm` enum is in place to plug in libigl BBW behind a future opt-in `-DENABLE_LIBIGL_BBW` flag for users who accept the GPL implications.

## Algorithm

For each vertex `v`:

1. For every bone `b` in the skeleton's bind pose, compute distance from `v` to the bone's segment (line from head to average of children, falling back to point distance for leaves).
2. If `dist > maxInfluenceDistance × mesh-diagonal`, skip bone (prevents a finger from picking up weight on a foot).
3. Compute raw weight `w_b = 1 / (dist² + eps)^(falloff/2)`.
4. Keep the top-K bones (default K=4 matches the hardware skinning convention).
5. Normalize so kept weights sum to 1.

This is the same algorithm Maya / 3dsMax use as their default "smooth bind." It's heuristic, not biharmonic, so it doesn't get BBW's volumetric smoothness — but it works out of the box on any character mesh including non-manifold FBX imports.

`replaceExisting=false` enables a merge mode that fills in missing weights while keeping existing ones (useful for "manually skinned the torso, now auto-skin the rest" workflows).

## Surface

- **CLI**: `qtmesh skin <file> [--max-influences N] [--falloff F] [--max-distance D] [--skip-unweighted] [--merge] -o <out> [--json]`. Multi-entity inputs rejected fail-fast (matches the `decimate` / `retopo` convention).
- **MCP**: `compute_skin_weights` tool with `max_influences / falloff / max_distance / skip_unweighted / replace_existing` params.
- **GUI**: Edit Mode → Tools → "Compute Skin Weights…" button + top-level dialog (`qml/SkinWeightsDialog.qml`) driven by the `SkinWeightsController` QML singleton. Button binds to `hasSkinnedSelection` so it disables on static meshes.
- **Library**: `SkinWeights::computeAndApply(entity, opts, algo)` for the Ogre-backed path; `SkinWeights::computeWeights(positions, bones, opts, outWeights)` is a pure-data variant used by tests.

## Verification (Rumba Dancing.fbx)

```
$ qtmesh skin 'rumba115/Rumba Dancing.fbx' -o /tmp/reskin.glb
Skin Weights
============
Mesh:               Rumba Dancing
Skeleton:           Rumba Dancing.skeleton
Bones:              69
Vertices processed: 5828
Assignments:        8461 → 20129
Wrote: reskin.glb
```

20,129 assignments / 5828 verts = avg 3.45 influences per vertex (capped at 4). Valid 761 KB .glb produced.

## Sentry breadcrumbs

- `ai.assist.skin_weights` for every action (CLI, MCP, GUI).

## Tests

`src/SkinWeights_test.cpp` covers the pure-data path:

- Near vertices get correct bone assignment
- Weights sum to 1.0 per vertex
- All weights non-negative
- `max_influences` cap respected
- `max_distance` cap excludes far bones
- Higher falloff sharpens the bind toward the nearest bone
- Empty input returns false
- Algorithm string round-trip

## Test plan

- [x] `qtmesh skin <file> -o <out>` produces valid skinned output
- [x] `--max-influences N` clamps assignments per vertex (verified: 5828 verts × 2 = 10,596 assignments with `--max-influences 2`)
- [x] Invalid numeric flags rejected with clear error
- [x] `--json` emits structured output
- [x] Multi-entity input rejected fail-fast
- [x] GUI button binds to skinned selection only
- [ ] CI green on Linux / macOS / Windows
- [ ] CodeRabbit / Sonar review

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Automatic skin-weight computation for skinned meshes (inverse-distance heuristic)
  * "Compute Skin Weights…" dialog in Animation Mode → Mode Tools with configurable parameters and keyboard shortcuts; button disabled when no skinned selection
  * Headless CLI and MCP tool endpoints for batch/remote skin-weight processing
  * Undo/redo support for computed weight changes; JSON/text reporting of results

* **Documentation**
  * CLI and README updated with new skin-weight command options

* **Tests**
  * Unit tests covering core algorithm behavior and reporting

---

## Follow-ups (commits 2910c19, bc4949c)

- **Undo/redo**: the auto-skin now runs through `ComputeSkinWeightsCommand` — snapshots the bone-assignment lists before, restores + recompiles on undo. Ctrl+Z reverts a bad auto-skin, Ctrl+Shift+Z reapplies it. A skeleton pre-check avoids leaving a no-op undo entry.
- **Button moved to Animation Mode**: skinning governs animation deformation (a rigging step, not a mesh edit), so "Compute Skin Weights…" now lives in a "Skinning" section under **Animation Mode → Mode Tools** rather than Edit Mode.
- **CodeRabbit review addressed** (2910c19): critical null-deref guard, controller error-string, MCP arg-type validation, CLI entity filter + import/export breadcrumbs, dialog keyboard accessibility, doc/test polish. All threads confirmed resolved by CodeRabbit.
- **Test coverage**: `SkinWeights_test` expanded 8 → 17 tests (edge cases + report serializers).

### CODE REVIEW

File: src/SkinWeights.cpp [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+            std::clamp(opts.maxInfluencesPerVertex, 1, 8));
+        for (size_t v = 0; v < weights.size(); ++v) {
+            const auto& vw = weights[v];
+            if (vw.count == maxK) ++subReport.verticesWithMaxInfluences;
+            for (int k = 0; k < vw.count; ++k) {
+                Ogre::VertexBoneAssignment vba;
+                vba.vertexIndex = static_cast<unsigned int>(v);
+                vba.boneIndex   = boneIdxToHandle[vw.boneIndices[k]];
+                vba.weight      = static_cast<float>(vw.weights[k]);
+                sub->addBoneAssignment(vba);
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/SkinWeights.cpp]: Fixed in dd890b8. Shared-vertex submeshes now route their bone assignments through `Mesh::clearBoneAssignments`/`addBoneAssignment`/`_compileBoneAssignments` (the list the FBX/glTF exporters read), and the shared vertex data is processed exactly once rather than per-submesh. Per-submesh (non-shared) data continues to route to the SubMesh APIs. Both paths share a single `computeAndCommit` lambda parameterized by the four owner-specific operations.

File: src/SkinWeights.cpp [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+
+        // Run the pure-data weight computation.
+        std::vector<SkinWeights::VertexWeights> weights;
+        SkinWeights::computeWeights(positions.data(),
+                                     static_cast<int>(vd->vertexCount),
+                                     bones, opts, weights);
+
+        if (opts.replaceExisting) {
+            sub->clearBoneAssignments();
+        }
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/SkinWeights.cpp:313]: Fixed in dd890b8. In merge mode (`replaceExisting=false`) the emit loop now builds the set of already-weighted vertex indices from the existing assignment list up front and skips emitting new weights for those vertices — filling only the unweighted ones. Verified on Rumba Dancing.fbx (fully pre-skinned): merge mode now reports 8461 → 8461 assignments (no-op) instead of doubling to ~28590.

File: src/SkinWeightsController.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            .arg(maxInfluencesPerVertex)
+            .arg(falloff).arg(maxInfluenceDistance));
+
+    m_busy = true;
+    emit busyChanged();
+
+    SkinWeightsReport report;
+    try {
+        report = SkinWeights::computeAndApply(entity, opts);
+    } catch (const Ogre::Exception& e) {
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/SkinWeightsController.cpp:135]: Fixed in 2910c19 — added a `!entity || !entity->getMesh()` guard before dereferencing, returning a user-facing error instead of crashing.

File: src/SkinWeightsController.cpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    result["totalBones"]            = report.totalBones;
+    result["totalVerticesProcessed"] = report.totalVerticesProcessed;
+    result["totalAssignmentsBefore"] = report.totalAssignmentsBefore;
+    result["totalAssignmentsAfter"]  = report.totalAssignmentsAfter;
+    if (!report.error.isEmpty()) result["error"] = report.error;
+
+    if (report.applied) emit weightsApplied(result);
+    else                emit error(report.error.isEmpty()
+                                    ? QStringLiteral("Skin weights failed")
+                                    : report.error);
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/SkinWeightsController.cpp]: Fixed in 2910c19 — the generic failure path now writes the fallback message into result["error"] before emitting, so the dialog no longer overwrites it with "unknown error".

File: src/MCPServer.cpp [hunk 5]
```diff
@@ -1478,6 +1480,62 @@ QJsonObject MCPServer::toolRetopologize(const QJsonObject &args)
...
+    if (args.contains("max_influences"))
+        opts.maxInfluencesPerVertex = args["max_influences"].toInt(4);
+    if (args.contains("falloff"))
+        opts.falloff = args["falloff"].toDouble(4.0);
+    if (args.contains("max_distance"))
+        opts.maxInfluenceDistance = args["max_distance"].toDouble(0.5);
+    if (args.contains("skip_unweighted"))
+        opts.skipUnweightedBones = args["skip_unweighted"].toBool(false);
+    if (args.contains("replace_existing"))
+        opts.replaceExisting = args["replace_existing"].toBool(true);
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/MCPServer.cpp:1516]: Fixed in 2910c19 — added isDouble()/isBool() type checks that return a usage error before reading any arg, so string-typed numbers/bools no longer silently coerce to defaults.

File: src/CLIPipeline.cpp [hunk 6]
```diff
@@ -6866,6 +6873,125 @@ int CLIPipeline::cmdRetopo(int argc, char* argv[])
...
+        err() << "Error: skin weights failed — " << report.error << Qt::endl;
+        return 1;
+    }
+
+    auto* node = entity->getParentSceneNode();
+    const QString fmt = formatForExtension(outputPath);
+    if (MeshImporterExporter::exporter(node, QFileInfo(outputPath).absoluteFilePath(), fmt) != 0) {
+        err() << "Error: export failed." << Qt::endl;
+        return 1;
+    }
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/CLIPipeline.cpp:6994]: Fixed in 2910c19 — added file.import and file.export breadcrumbs around the I/O steps in cmdSkin.

File: src/CLIPipeline.cpp [hunk 7]
```diff
@@ -6866,6 +6873,125 @@ int CLIPipeline::cmdRetopo(int argc, char* argv[])
...
+        err() << "Error: failed to load " << inputPath << Qt::endl; return 1;
+    }
+    if (entities.size() > 1) {
+        err() << "Error: " << inputPath
+              << " contains multiple mesh entities. `qtmesh skin` "
+                 "currently supports one entity per file."
+              << Qt::endl;
+        return 1;
+    }
+    Ogre::Entity* entity = entities.first();
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/CLIPipeline.cpp]: Fixed in 2910c19 — cmdSkin now filters Manager::getEntities() on getMovableType() == "Entity" before the count/first checks.

File: qml/SkinWeightsDialog.qml [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+                }
+            }
+            InspectorLabel { text: parent.parent.label }
+        }
+        MouseArea {
+            anchors.fill: parent
+            cursorShape: Qt.PointingHandCursor
+            onClicked: parent.toggled()
+        }
+    }
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply qml/SkinWeightsDialog.qml:200]: Fixed in 2910c19 — InspectorButton/InspectorCheckbox now set activeFocusOnTab, Accessible role/name/checked, Enter/Space/Return handlers, plus a focus-ring border.

File: src/SkinWeights.h [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    // is more than `maxInfluenceDistance * (mesh bounding-box
+    // diagonal)` is excluded. Stops a finger bone from getting any
+    // weight on a foot vertex. 0 disables the cap.
+    double maxInfluenceDistance = 0.5;
+
+    // Skip bones with zero existing vertex weights when reading
+    // back results? Mixamo skeletons ship dozens of helper bones
+    // (twist bones, IK targets) that aren't actually skinned.
+    // When true, those bones are ignored. Default true.
+    bool skipUnweightedBones = false;
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/SkinWeights.h:71]: Fixed in 2910c19 — corrected the comment to "Default false".

File: CLAUDE.md [hunk 10]
```diff
@@ -237,6 +237,7 @@ Three singletons manage core state. All run on the main thread. Access via `Clas
...
...ent (line from bone-head to the average of its children, falling back to point distance for leaf bones in the skeleton's bind pose), apply `1/dist^falloff` weighting, keep the top-K bones (default K=4 matches hardware skinning), and normalize. This is the same algorithm Maya / 3dsMax use as their default "smooth bind." Distance cap (`maxInfluenceDistance` × mesh-diagonal) prevents a finger bone from picking up weight on a foot. Optional `skipUnweightedBones` filters Mixamo helper bones. `replaceExisting=false` enables a merge mode for "fill in missing weights" workflows. Surfaced via `qtmesh skin --max-influences N --falloff F -o out`, MCP `compute_skin_weights`, and the **Edit Mode → Tools → "Compute Skin Weights…" button** (`qml/SkinWeightsDialog.qml`, driven by `SkinWeightsController` singleton). The button binds to `hasSkinnedSelection` so it disables on static (skeleton-less) meshes. Sentry breadcrumb category `ai.assist.skin_weights`. A future slice can plug libigl BBW in behind `-DENABLE_LIBIGL_BBW` for users who accept the GPL implications. Verified on Rumba Dancing.fbx: 69 bones, 5828 verts → 20,129 vertex-bone assignments (avg 3.45 influences/vert), valid glTF round-trip.
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply CLAUDE.md]: Fixed in 2910c19 — added retopo and skin to the recognized-subcommand list.

File: src/SkinWeights_test.cpp [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+    SkinWeightsOptions optsHigh = optsLow;
+    optsHigh.falloff = 8.0;
+    SkinWeights::computeWeights(kBarPositions.data(), 4, kTwoBones, optsHigh, high);
+
+    // Vertex 1 sits at y=1 — equidistant-ish from both bones.
+    // High falloff should drive its primary weight higher.
+    ASSERT_GE(low[1].count, 1);
+    ASSERT_GE(high[1].count, 1);
+    EXPECT_GE(high[1].weights[0], low[1].weights[0])
+        << "high falloff failed to concentrate weight on the nearest bone";
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/SkinWeights_test.cpp]: Fixed in 2910c19 — added ASSERT_TRUE(computeWeights(...)) and size asserts before indexing.

File: qml/PropertiesPanel.qml [hunk 12]
```diff
@@ -301,6 +301,22 @@ Rectangle {
...
+            // Issue #402: auto skin weights. Surfaced in Animation
+            // Mode because skinning governs how the mesh deforms
+            // under animation — it's a rigging step, not a mesh-topology
+            // edit. Only meaningful on a skinned (skeleton-bearing)
+            // mesh; the button inside disables itself otherwise.
+            CollapsibleSection {
+                title: "Skinning"
+                sectionVisible: root.modeToolSectionVisible(
+                    EditorModeController.AnimationMode,
+                    PropertiesPanelController.hasAnimations)
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply qml/PropertiesPanel.qml]: Fixed in 9c14ac7 — the Skinning section is now gated on SkinWeightsController.hasSkinnedSelection (a skeleton) plus the Animation-Mode + Mode-Tools-tab check, not hasAnimations. It now shows for skinned meshes with no clips yet.

File: qml/PropertiesPanel.qml [hunk 13]
```diff
@@ -1092,6 +1108,64 @@ Rectangle {
...
+                    enabled: SkinWeightsController.hasSkinnedSelection
+                    cursorShape: SkinWeightsController.hasSkinnedSelection
+                        ? Qt.PointingHandCursor : Qt.ForbiddenCursor
+                    onClicked: root.openSkinWeightsDialog()
+                    ToolTip.visible: containsMouse
+                    ToolTip.delay: 500
+                    ToolTip.text: SkinWeightsController.hasSkinnedSelection
+                        ? "Compute per-vertex bone weights via inverse-distance to bone segments. Mesh must have a skeleton."
+                        : "Select a skinned mesh (with a skeleton) first."
+                }
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply qml/PropertiesPanel.qml:1178]: Fixed in 9c14ac7 — added activeFocusOnTab, Accessible.role/name, Enter/Space/Return handlers, and a focus-ring border to the Compute Skin Weights button (matching the dialog controls).

File: src/commands/ComputeSkinWeightsCommand.cpp [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+
+Ogre::Entity* ComputeSkinWeightsCommand::resolveEntity() const
+{
+    Manager* mgr = Manager::getSingletonPtr();
+    if (!mgr) return nullptr;
+    for (Ogre::Entity* e : mgr->getEntities()) {
+        if (e && e->getMovableType() == "Entity"
+            && e->getName() == mEntityName)
+            return e;
+    }
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/commands/ComputeSkinWeightsCommand.cpp:33]: No behavior change needed — Manager::getEntities() already filters to real Ogre::Entity objects at collection time (collectEntitiesRecursive only appends attached objects whose getMovableType()=="Entity"), so the list never contains ManualObjects. Added a comment in 9c14ac7 documenting that the inner re-check is belt-and-suspenders, not load-bearing.

File: src/SkinWeightsController.cpp [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+    // Pre-check the skeleton here so a skeleton-less mesh fails
+    // cleanly WITHOUT leaving a no-op entry on the undo stack
+    // (QUndoStack::push runs the command's redo() unconditionally).
+    if (!entity->getMesh()->getSkeleton()) {
+        const auto msg = QStringLiteral("Mesh has no skeleton attached.");
+        emit error(msg);
+        result["applied"] = false;
+        result["error"]   = msg;
+        return result;
+    }
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/SkinWeightsController.cpp:105]: Fixed in 9c14ac7 — added a ui.action breadcrumb at the top of computeWeightsForSelected so failed attempts (no selection / no skeleton / stale entity) reach Sentry before the validation early-returns.
